magenta-lib/src/test/scala/magenta/deployment_type/CloudFormationTest.scala

```diff
@@ -119,11 +119,118 @@ class CloudFormationTest extends FlatSpec with Matchers with Inside with EitherV
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = false), TemplateParameter("param3", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map("param3" -> "user-provided"))
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+      "param3" -> SpecifiedValue("user-provided")
+    ))
+  }
+
+  it should "always specify parameters when the user directly supplies values" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "cheese", None), ExistingParameter("param3", "something-else", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = false), TemplateParameter("param3", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map("param3" -> "user-provided"))
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+      "param3" -> SpecifiedValue("user-provided")
+    ))
+  }
+
+  it should "always specify parameters when the template contains deploy parameters" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("Stack", "cfn", None), ExistingParameter("Stage", "CODE", None), ExistingParameter("BuildId", "540", None))
+    val templateParameters =
+      List(TemplateParameter("Stack", default = true), TemplateParameter("Stage", default = false), TemplateParameter("BuildId", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map.empty)
+
+    combined.right.value should be(Map(
+      "Stack" -> SpecifiedValue("cfn"),
+      "Stage" -> SpecifiedValue("PROD"),
+      "BuildId" -> SpecifiedValue("543")
+    ))
+  }
+
+  it should "fail when there are new parameters in the template that don't have default values" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "monkey", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = false), TemplateParameter("param2", default = false), TemplateParameter("param3", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map.empty)
+
+    combined.left.value should startWith("Missing parameters for param3:")
+  }
+
+  it should "fail when the user provides parameters that are not in the template" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
```

Comment:
The user here being the contents of `riff-raff.yaml`?

Reply:
Yeah - either the `templateParameters` directly or the result of the `amiParameters`/`amiParametersToTags` params in the `riff-raff.yaml`.